tests/e2e/e2e_test.go

```diff
@@ -1692,3 +1695,65 @@ func (s *IntegrationTestSuite) TestAConcentratedLiquidity_CanonicalPool_And_Para
 	s.Require().NoError(err)
 	osmoassert.DecApproxEq(s.T(), expectedSpotPrice, concentratedPool.GetCurrentSqrtPrice().Power(2), sdk.NewDecWithPrec(1, 3))
 }
+
+// TestPoolMigration ensures that message `UnlockAndMigrateSharesToFullRangeConcentratedPosition` performs as expected by
+// testing four cases (superfluid bonded, superfluid unbonding, unlocked) with each cases testing and ensuring the following upon the message being called:
+// - ensure that existing delegation amount post migration is as expected
+// - ensure that new delegation post migration in the intermediary account is as expected
+// - ensure that the amount locked in old lock post migration is as expected
+// - ensure that the amount locked in the new lock that has been created from lock is as expected.
+func (s *IntegrationTestSuite) TestPoolMigration() {
+	chain := s.configurer.GetChainConfig(0)
+	node, err := chain.GetDefaultNode()
+	s.NoError(err)
+
+	percentOfSharesToMigrate := sdk.MustNewDecFromStr("0.9")
+	fundTokens := []string{"50000000000uosmo", "50000000000stake"}
+	poolJoinAddress := node.CreateWalletAndFund("poolJoinAddress", fundTokens)
+
+	// Get the permissionless pool creation parameter.
+	isPermissionlessCreationEnabledStr := node.QueryParams(cltypes.ModuleName, string(cltypes.KeyIsPermisionlessPoolCreationEnabled))
+	if !strings.EqualFold(isPermissionlessCreationEnabledStr, "true") {
+		// Change the parameter to enable permissionless pool creation.
+		err = chain.SubmitParamChangeProposal("concentratedliquidity", string(cltypes.KeyIsPermisionlessPoolCreationEnabled), []byte("true"))
+		s.Require().NoError(err)
+
+		// Confirm that the parameter has been changed.
+		isPermissionlessCreationEnabledStr = node.QueryParams(cltypes.ModuleName, string(cltypes.KeyIsPermisionlessPoolCreationEnabled))
+		if !strings.EqualFold(isPermissionlessCreationEnabledStr, "true") {
+			s.T().Fatal("concentrated liquidity pool creation is not enabled")
+		}
+	}
+
+	tokenOutMins := sdk.NewCoins(sdk.NewCoin("uosmo", sdk.NewInt(200000)), sdk.NewCoin("stake", sdk.NewInt(200000)))
+
+	// Case 1: SuperfluidBonded
+	var (
+		superfluidDelegated    = true
+		superfluidUndelegating = false
+		unlocking              = false
+		noLock                 = false
+	)
+	s.testPoolMigration(chain, poolJoinAddress, superfluidDelegated, superfluidUndelegating, unlocking, noLock, percentOfSharesToMigrate, tokenOutMins)
+
+	// Case 2: SuperfluidUnbonding
+	superfluidDelegated = true
+	superfluidUndelegating = true
+	unlocking = true
+	noLock = false
+	s.testPoolMigration(chain, poolJoinAddress, superfluidDelegated, superfluidUndelegating, unlocking, noLock, percentOfSharesToMigrate, tokenOutMins)
+
+	// Case 3: SuperfluidUnbonding
+	superfluidDelegated = false
+	superfluidUndelegating = false
+	unlocking = true
+	noLock = false
+	s.testPoolMigration(chain, poolJoinAddress, superfluidDelegated, superfluidUndelegating, unlocking, noLock, percentOfSharesToMigrate, tokenOutMins)
```

Comment:
what are the differences between these two cases?

Reply:
Added @czarcas7ic